Sprint #13

JudjinGM opened the pull request:
This is pull request for review and merge my homework for Sprint #13.

JudjinGM committed:
Spring 13. New audio player screen.

JudjinGM committed:
Sprint 13. New audio player screen.

JudjinGM committed:
Sprint 13. New audio player screen.

JudjinGM committed:
Sprint 13. New audio player screen.

JudjinGM committed:
Sprint 13. Refactoring #1

JudjinGM committed:
Sprint 13. Refactoring #2

kirillNay commented on app/src/main/java/com/example/playlistmaker/ui/AudioPlayerActivity.kt:
```diff
@@ -0,0 +1,81 @@
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_audioplayer)
+
+        track = Gson().fromJson(intent.getStringExtra(TRACK), Track::class.java)
```
⏫ Обычно, когда возникает необходимость передавать некий объект через `Intent`, сохранять объект в `onSavedInstanceState` или сохранять объект в аргументах `Fragment`'ов (про фрагменты будет рассказано в следующих спринтах) андроид использует `Bundle` для сохранения и передачи данных.
Особенность `Bundle` заключается в том, что он представляет собой коллекцию пар _"ключ-значение"_, где ключи являются строками, а значения могут быть различных типов данных, таких как строки, числа, объекты `Parcelable` и `Serializable`, то есть сериализуемые объекты.

В Android существует несколько способов сериализации объектов:

- `Serializable` - это интерфейс, который позволяет сериализовать объекты и передавать их между компонентами Android. Для того, чтобы использовать `Serializable`, класс должен реализовывать этот интерфейс.

- `Parcelable` - это интерфейс, который также позволяет сериализовать объекты, но он оптимизирован для передачи между компонентами Android. Для того, чтобы использовать `Parcelable`, класс должен реализовывать этот интерфейс, а затем объект можно передать через `Intent` или сохранить в `Bundle`.

- `JSON` - это формат данных, который чаще используется для передачи данных между клиентом и сервером. В Android можно использовать библиотеки, такие как GSON или Moshi, для сериализации и десериализации объектов в формате JSON.

Таким образом, если мы хотим передавать данные внутри `Bundle` нам лучше использовать именно `Parcelable` для более оптимальной работы. 

P.S. нашел в интернетах такой график, который сравнивает разные версии андроидовских JVM. Но все же на нем можно увидеть, насколько использование `Parcelable` является быстрым по сравнению с `GSON`. 
![image](https://user-images.githubusercontent.com/56832972/232862116-0fe439e5-1770-4bce-837f-89ddbc53aa17.png)

kirillNay commented on app/src/main/res/drawable-night/library_button.xml:
```diff
@@ -0,0 +1,13 @@
+    android:viewportHeight="51">
+  <path
+      android:pathData="M25.5,0L25.5,0A25.5,25.5 0,0 1,51 25.5L51,25.5A25.5,25.5 0,0 1,25.5 51L25.5,51A25.5,25.5 0,0 1,0 25.5L0,25.5A25.5,25.5 0,0 1,25.5 0z"
+      android:fillColor="#ffffff"
```
⏫ Цвета в векторных изображениях также можно выносить в файл ресурсов. Так, можно организовать более гибкую дизайн-систему приложения.
Кроме того, это поможет экономить занимаемое ресурсами место, если нам нужно иметь иконки разных цветов для разных тем приложения. Ведь вместо того, чтобы создавать две иконки и класть в drawable и `drawable-night`, мы можем создать одну иконку в `drawable`, а файл ресурса с цветом будет задан в `values` и `values-night`

kirillNay commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -5,14 +5,15 @@
         app:layout_constraintStart_toStartOf="parent"
-        app:layout_constraintTop_toTopOf="parent" />
+        app:layout_constraintTop_toTopOf="parent"
+        android:contentDescription="@string/back_button_image" />
```
👍 Отличная идея добавлять contentDescription к кнопкам!

⏫ Также чтобы повысить `Accessibility` можно динамически устанавливать `contentDescription` у элементов, например у `TextView` с информацией о треке

JudjinGM committed:
Sprint 13. Refactoring #3
Track now Parcelable, also changed default fields for null safety to use default Track instance in AudioPlayerActivity.kt while retrieving instance from intent.
